lib/spack/spack/cmd/logs.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import gzip
+import os
+import shutil
+import sys
+
+import spack.cmd
+import spack.util.compression as compression
+from spack.cmd.common import arguments
+
+description = "print out logs for packages"
+section = "basic"
+level = "long"
+
+
+def setup_parser(subparser):
+    arguments.add_common_arguments(subparser, ["spec"])
+
+
+def _dump_byte_stream_to_stdout(instream):
+    if sys.platform == "win32":
+        # https://docs.python.org/3/library/sys.html#sys.stdout unconditionally
+        # recommends using .buffer to write binary data to sys.stdout, but
+        # https://docs.python.org/3/library/io.html#io.TextIOBase.buffer claims
+        # it is not available in all implementations, so only try accessing it
+        # on Windows, where stdout.fileno() is not useful
+        try:
+            outstream = sys.stdout.buffer
+        except AttributeError:
+            raise spack.main.SpackCommandError(
+                "This command must be invoked in a"
+                " context where stdout is accessible"
+                f" as a binary stream (stdout = {type(sys.stdout)})"
+            )
+    else:
+        outstream = os.fdopen(sys.stdout.fileno(), "wb", closefd=False)
+
+    shutil.copyfileobj(instream, outstream)
+
+
+def dump_build_log(package):
+    with open(package.log_path, "rb") as f:
+        _dump_byte_stream_to_stdout(f)
+
+
+def _logs(cmdline_spec, concrete_spec):
+    if concrete_spec.installed:
+        log_path = concrete_spec.package.install_log_path
+    elif os.path.exists(concrete_spec.package.stage.path):
+        dump_build_log(concrete_spec.package)
+        return
+    else:
+        raise spack.main.SpackCommandError(f"{cmdline_spec} is not installed or staged")
+
+    if not os.path.exists(log_path):
+        raise spack.main.SpackCommandError(f"No logs are available for {cmdline_spec}")
+
+    compression_ext = compression.extension_from_file(log_path)
```

Comment:
`extension_from_file` is doing a lot of unnecessary things, and it's a complete misnomer. Can you just branch on `Path(p).suffix`?

Reply:
> it's a complete misnomer

It's definitely not, the test added in this PR (among all the original tests for the function itself) ensure that: I create a `.gz` file, and this function identifies it as a `gz` (otherwise the test would fail trying to read compressed output).

> is doing a lot of unnecessary things,

we have a library function that tells us the extension of a file, I don't think it's the business of this code to worry about the implementation details of that function

>See also nitpick below about `open`, you shouldn't open or stat the file more than once.

As I mention there, this function has no guarantees about the ongoing existence of files (including after we've opened them), so I don't think this is a problem.

Reply:
Can you please look at the implementation? :)

Reply:
> It's not deriving an "extension" from a file name,

I never said it did. My point is that I don't care how it derives the extension: I give it a file name, it gives me the extension.

> cause otherwise you're interacting with the file on the file system yet a third time, which can raise exceptions, which you don't handle.

A recurring theme of your comments seems to be a concern about the possibility of the state of the file changing between my checks. This command has no control over that because the files in question can be deleted by another process at any time. I don't handle those exceptions because there is no recourse for handling them and they should in fact terminate the process.

Also, the fact that the file might be deleted in this scenario

```
ext = compression.extension_from_file(path)
# the file was deleted
if ext == .gz
  gzip.open
```

is the same as this scenario

```
ext = path.split()[-1] # or whatever other non-fs-interacting mechanism you intend to use to determine the extension
# The file was deleted right here
if ext == .gz
  gzip.open
```

I don't care in the first case as much as I don't care in the second case: I don't depend on the continued existence of the file between determining the extension and trying to open it.

Reply:
It is not the same scenario if you take my advice into account.

```python
try:
  handle = open(path, "rb")
except OSError:
  # deal with non-existence / permission errors once and only once.
with maybe_gzipped(handle) as f:
  do_something(f)
```

there should not be stat or open calls above `try`, so no `extension_from_file` calls, opening once is enough.

Reply:
One new implied desire in https://github.com/spack/spack/pull/42202#discussion_r1470232557 is for better error messages. Up to this point, my code had no `except` in it, so I will argue that this is all effectively the same in that case. However, if `except` is used, then paying attention to the possible errors of `compression.extension_from_file` makes sense: [4d44aec](https://github.com/spack/spack/pull/42202/commits/4d44aec8361650cbbd9fd7d595224e72fe27f97a) achieves that by wrapping the extension detection and `open()` into a single `try` block (and updates the tests for that).